subprojects/test-runner/client/src/main/kotlin/com/avito/runner/scheduler/metrics/TestMetricsAggregatorImpl.kt

```diff
@@ -1,66 +1,70 @@
 package com.avito.runner.scheduler.metrics
 
+import com.avito.android.Result
 import com.avito.math.median
 import com.avito.runner.scheduler.metrics.model.DeviceKey
 import com.avito.runner.scheduler.metrics.model.DeviceTimestamps
+import com.avito.runner.scheduler.metrics.model.TestTimestamps
 
-internal class TestMetricsAggregatorImpl(
+internal data class TestMetricsAggregatorImpl(
     private val testSuiteStartedTime: Long,
     private val testSuiteEndedTime: Long,
     private val deviceTimestamps: Map<DeviceKey, DeviceTimestamps>
 ) : TestMetricsAggregator {
 
     private val testTimestamps = deviceTimestamps.flatMap { it.value.testTimestamps.values }
 
-    private val firstTestStarted = testTimestamps
-        .mapNotNull { it.started }
+    private val firstTestStarted: Result<Long> = testTimestamps.filterIsInstance<TestTimestamps.Finished>()
+        .map { it.startTime }
         .minOrNull()
+        .toResult { "Cannot calculate first started test time" }
 
-    private val lastTestEnded = testTimestamps
-        .mapNotNull { it.finished }
+    private val lastTestEnded: Result<Long> = testTimestamps.filterIsInstance<TestTimestamps.Finished>()
+        .map { it.finishTime }
         .maxOrNull()
+        .toResult { "Cannot calculate last ended test time" }
 
-    private val queueTimes: List<Long> = testTimestamps
-        .mapNotNull { it.onDevice }
-        .map { it - testSuiteStartedTime }
+    private val queueTimes: List<Long> = testTimestamps.filterIsInstance<TestTimestamps.Finished>()
+        .map { it.onDevice - testSuiteStartedTime }
 
-    private val installationTimes: List<Long> = testTimestamps
-        .mapNotNull { it.installationTime }
+    private val installationTimes: List<Long> = testTimestamps.filterIsInstance<TestTimestamps.Finished>()
+        .map { it.installationTime }
 
-    override fun initialDelay(): Long? = firstTestStarted?.let { it - testSuiteStartedTime }
+    override fun initialDelay(): Result<Long> = firstTestStarted.map { it - testSuiteStartedTime }
 
-    override fun endDelay(): Long? = lastTestEnded?.let { testSuiteEndedTime - it }
+    override fun endDelay(): Result<Long> = lastTestEnded.map { testSuiteEndedTime - it }
 
-    override fun medianQueueTime(): Long? = queueTimes.aggregateOrNull { it.median() }
+    override fun medianQueueTime(): Result<Long> = queueTimes.aggregate(
+        { it.median() },
+        { "Cannot calculate median queue time" }
```

Comment:
It'd be good to help a bit in troubleshooting. 
"Cannot do something" is the consequence. If I see this error, I want to understand why.
Maybe dumping related events will help here. Up to you.